k8s/admission.go

```diff
@@ -0,0 +1,166 @@
+	ErrReasonFieldNotAllowed = "field_not_allowed"
+
+	errStringNoAdmissionControllerDefined = "no %s admission controller defined for group '%s' and kind '%s'"
+)
+
+// SimpleAdmissionError implements resource.AdmissionError
+type SimpleAdmissionError struct {
+	error
+	statusCode int
+	reason     string
+}
+
+// StatusCode returns the error's HTTP status code
+func (s *SimpleAdmissionError) StatusCode() int {
+	return s.statusCode
+}
+
+// Reason returns a machine-readable reason for the error
+func (s *SimpleAdmissionError) Reason() string {
+	return s.reason
+}
+
+// NewAdmissionError returns a new SimpleAdmissionError, which implements resource.AdmissionError
+func NewAdmissionError(err error, statusCode int, reason string) *SimpleAdmissionError {
+	return &SimpleAdmissionError{
+		error:      err,
+		statusCode: statusCode,
+		reason:     reason,
+	}
+}
+
+// OpinionatedMutatingAdmissionController is a MutatingAdmissionController which wraps an optional user-defined
+// Mutate() function with a set of additional PatchOperations which set metadata and label properties.
+type OpinionatedMutatingAdmissionController struct {
+	Underlying resource.MutatingAdmissionController
+}
+
+// now is used to wrap time.Now so it can be altered for testing
+var now = time.Now
+
+// Mutate runs the underlying MutateFunc() function (if non-nil), and if that returns successfully,
+// appends additional patch operations to the MutatingResponse for CommonMetadata fields not in kubernetes standard metadata,
+// and labels internally used by the SDK, such as the stored version.
+func (o *OpinionatedMutatingAdmissionController) Mutate(ctx context.Context, request *resource.AdmissionRequest) (*resource.MutatingResponse, error) {
+	// Get the response from the underlying controller, if it exists
+	var err error
+	var resp *resource.MutatingResponse
+	if o.Underlying != nil {
+		resp, err = o.Underlying.Mutate(ctx, request)
+		if err != nil {
+			return resp, err
+		}
+	}
+	if resp == nil || resp.UpdatedObject == nil {
+		resp = &resource.MutatingResponse{
+			UpdatedObject: request.Object,
+		}
+	}
+
+	// Get the CommonMetadata, so we can update it
+	cmd := resp.UpdatedObject.CommonMetadata()
+
+	if cmd.Labels == nil {
+		cmd.Labels = make(map[string]string)
+	}
+	cmd.Labels[versionLabel] = request.Version
+
+	// Operation-based changes
+	switch request.Action {
+	case resource.AdmissionActionCreate:
+		cmd.CreatedBy = request.UserInfo.Username
+		cmd.UpdateTimestamp = cmd.CreationTimestamp
+	case resource.AdmissionActionUpdate:
+		cmd.UpdatedBy = request.UserInfo.Username
+		cmd.UpdateTimestamp = now()
+	default:
+		// Do nothing
+	}
+	resp.UpdatedObject.SetCommonMetadata(cmd)
+	return resp, nil
+}
+
+// NewOpinionatedMutatingAdmissionController creates a pointer to a new OpinionatedMutatingAdmissionController wrapping the
+// provided mutateFunc (nil mutateFunc argument is allowed, and will cause the controller to not call the underlying function)
```

Comment:
Looks like the parameter here is no longer `mutateFunc`. I understand that its coming from `wrap.MutateFunc`? The comment reads like it was written for an older form of this function.

Reply:
Ah, good catch, let me update the docs